Contributions/Linux_Memory_Management_Essentials.md

```diff
@@ -0,0 +1,167 @@
+
+
+## **Disclaimer**
+* This document is not intended to be a replacement for understanding the memory management of the Linux,
+  nor it attempts to be an exhaustive analysis of safety implications.
+* Because of the very volatile nature of the code within the Linux Kernel, each of the statements made
+  below should not be taken at face value, but rather verified, for any Linux kernel version following
+  the one used while writing the document.
+* When referring to specific HW features, the document refers to the ARM64 architecture.
+
+## **Purpose of the document**
+This document aims to providing an holistic view of what happens in Linux memory management, so that
+one is at least aware of certain features and can use this document as jumping pad toward more detailed documentation.
+Or even to the code base.
+
+## **Structure of the document**
+The document is divided in two parts, based on hte destinatary of the memory allocations discussed: kernel-space or user-space.
+Individual points are numbered for ease of reference, but the numbering is not meant to represent any sequence.
+
+## **Memory management in Linux**
+
+### **Kernel-space memory allocations**
+
+#### **Facts**
+1. unlike processes memory, kernel memory pages are not swapped, nor dropped silently by the kernel itself,
+   although an hypervisor will do to a VM what the kernel does to a process (but this is beyond the control of the kernel)
+2. the kernel context (usually EL1 on ARM64) uses one single memory map (page tables) across all the cores
+   executing in kernel mode
+3. on 64 bit systems (e.g. ARM64 and x86_64), usually almost all physical memory is mapped in a
+   (semi)contiguous (there can be holes) range. Memory within this range is both virtually and physically contiguous.
+4. physically contiguous memory is treated as a scarce resource, and typically is not provided to userspace,
+   unless it explicitly asks for it (e.g. for implementing some DMA buffer)
+5. the kernel can access userspace in two ways:
+   1. through the userspace mapping
+      1. it is done in exceptional cases, like copy_to_user_copy_from_user
+      2. it uses the userspace memory map, which can implement HW protections against kernel misuses
+      3. it allows the kernel to see process pages in the same sequence and with the same mappings as the proces does
+   2. through the EL1 mappings + linear map
+      1. not the intended way to access process context
+      2. not very useful, even for an attacker, in a security scenario
+      3. the sequence of pages mapped in the userspace process is not known
+      4. some process pages might even be "missing", because they have been either swapped out or dropped
+      5. bypasses any protection that the process might employ through its own mapping
+6. attempt to keep pages hot in the system cache make it very difficult to predict which pages might surface in a certain place:
```

Comment:
* suface -> emerge, in the sense that one can look at the underlying memory management features and htnk that a certain page is acquired, released, and then it gets recycled elsewhere.
* the intent is to show how it is practically impossible to make any realistic assessment about the health of the memory subsystem, because of its intricacy and its exposure to a very large amount of fleeting perturbations, from both qualified and qm components, both from the kernel itself and user-space, driven by both internal and external events.

Reply:
Fixed in next version